In [1]:
# Basic imports
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import make_moons
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [2]:
# Q1(a)

X, y = make_moons(n_samples=500, noise=0.3, random_state=42)

print(X.shape)
print(y.shape)

(500, 2)
(500,)


In [3]:
# Q1(b)

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)

print(X_train.shape, X_test.shape)

(375, 2) (125, 2)


In [4]:
# Q2

from sklearn.tree import DecisionTreeClassifier

dt = DecisionTreeClassifier(random_state=42)

dt.fit(X_train, y_train)

y_pred = dt.predict(X_test)

acc_dt = accuracy_score(y_test, y_pred)

print("Decision Tree Accuracy:", acc_dt)

Decision Tree Accuracy: 0.856


### Q2 Answer

The Decision Tree classifier achieved an accuracy of **0.856** on the test set.

Yes, I expect the decision tree to overfit the dataset. Decision trees are highly flexible models and can easily grow very deep if no regularization is applied. This allows them to memorize the training data, which leads to overfitting and lower generalization performance on unseen data.


In [5]:
# Q3(a)

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier

log_clf = LogisticRegression()
rf_clf = RandomForestClassifier(random_state=42)
svc_clf = SVC(probability=True, random_state=42)
knn_clf = KNeighborsClassifier(n_neighbors=1)

voting_clf_hard = VotingClassifier(
    estimators=[
        ('lr', log_clf),
        ('rf', rf_clf),
        ('svc', svc_clf),
        ('knn', knn_clf)
    ],
    voting='hard'
)

voting_clf_hard.fit(X_train, y_train)

y_pred = voting_clf_hard.predict(X_test)

acc_hard = accuracy_score(y_test, y_pred)

print("Hard Voting Accuracy:", acc_hard)

Hard Voting Accuracy: 0.896


Hard Voting Accuracy: 0.896

In [6]:
# Q3(b)

voting_clf_soft = VotingClassifier(
    estimators=[
        ('lr', log_clf),
        ('rf', rf_clf),
        ('svc', svc_clf),
        ('knn', knn_clf)
    ],
    voting='soft'
)

voting_clf_soft.fit(X_train, y_train)

y_pred = voting_clf_soft.predict(X_test)

acc_soft = accuracy_score(y_test, y_pred)

print("Soft Voting Accuracy:", acc_soft)

Soft Voting Accuracy: 0.912


Soft Voting Accuracy: 0.912

### Q3(c) Answer

When comparing the results, the **soft voting classifier** achieved a higher accuracy (**0.912**) than the **hard voting classifier** (**0.896**).

This shows that soft voting performs better in this case.

The main reason is that hard voting only considers the final predicted class labels from each model and selects the majority vote. In contrast, soft voting takes into account the predicted probabilities of each class and averages them.

By using probabilities, soft voting can capture more information about the confidence of each model, leading to more accurate and balanced predictions.


In [7]:
# Q4

from sklearn.ensemble import BaggingClassifier

bag_clf = BaggingClassifier(
    DecisionTreeClassifier(),
    n_estimators=100,
    max_samples=100,
    bootstrap=True,
    random_state=42,
    n_jobs=-1
)

bag_clf.fit(X_train, y_train)

y_pred = bag_clf.predict(X_test)

acc_bag = accuracy_score(y_test, y_pred)

print("Bagging Accuracy:", acc_bag)

Bagging Accuracy: 0.92


### Q4 Answer

The Bagging classifier achieved an accuracy of **0.920**, which is significantly higher than the Decision Tree accuracy (**0.856**) obtained in Question 2.

This clearly shows that using a bagging ensemble improves performance compared to a single decision tree.

The improvement comes from the fact that bagging trains multiple decision trees on different bootstrap samples of the dataset and averages their predictions. This reduces the model's sensitivity to noise in the training data.

In terms of variance, using an ensemble model like bagging **decreases variance** compared to a single decision tree. This makes the model more stable and improves generalization performance.


In [8]:
# Q5(a)

rf = RandomForestClassifier(
    n_estimators=100,
    max_leaf_nodes=16,
    max_depth=None,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train, y_train)

y_pred = rf.predict(X_test)

acc_rf = accuracy_score(y_test, y_pred)

print("Random Forest Accuracy:", acc_rf)

Random Forest Accuracy: 0.928


Random Forest Accuracy: 0.928

In [9]:
# Q5(b)

print("Feature importances:", rf.feature_importances_)

Feature importances: [0.4237266 0.5762734]


Feature importances: [0.4237266 0.5762734]

### Q5(c) Answer

The Random Forest classifier achieved an accuracy of **0.928**, which is slightly higher than the Bagging classifier.

This indicates that Random Forest provides an improvement over standard bagging. The main reason is that Random Forest introduces additional randomness by selecting a random subset of features at each split. This reduces correlation between individual trees and leads to better generalization.

An advantage of Random Forest over Bagging is that it not only reduces variance but also decorrelates the trees, making the ensemble more robust and less prone to overfitting.

The feature importances obtained are **[0.4237, 0.5763]**, which shows that the second feature contributes more to the model’s predictions than the first feature.

Feature importance in Random Forest represents how much each feature contributes to reducing impurity (e.g., Gini impurity) across all trees in the forest.


In [10]:
# Q6(a)

from sklearn.ensemble import GradientBoostingClassifier

gb1 = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=2,
    random_state=42
)

gb1.fit(X_train, y_train)

y_pred = gb1.predict(X_test)

acc_gb1 = accuracy_score(y_test, y_pred)

print("GB (100, 0.1):", acc_gb1)

GB (100, 0.1): 0.896


GB (100, 0.1): 0.896

In [11]:
# Q6(b)

gb2 = GradientBoostingClassifier(
    n_estimators=300,
    learning_rate=0.01,
    max_depth=2,
    random_state=42
)

gb2.fit(X_train, y_train)

y_pred = gb2.predict(X_test)

acc_gb2 = accuracy_score(y_test, y_pred)

print("GB (300, 0.01):", acc_gb2)

GB (300, 0.01): 0.912


GB (300, 0.01): 0.912

### Q6(c) Answer

When comparing the two models, we observe that reducing the learning rate from **0.1** to **0.01** and increasing the number of estimators from **100** to **300** improved the accuracy from **0.896** to **0.912**.

This shows that a smaller learning rate can lead to better performance when it is compensated by using more trees. A lower learning rate makes the model learn more slowly and carefully, which helps improve generalization and reduces the risk of overfitting.

Boosting is a powerful technique because it builds models sequentially, where each new model focuses on correcting the errors made by the previous ones. This step-by-step error correction allows boosting methods to achieve high accuracy and strong predictive performance.


In [12]:
# Q7(a)

from sklearn.ensemble import StackingClassifier

stack_clf = StackingClassifier(
    estimators=[
        ('lr', log_clf),
        ('rf', rf_clf),
        ('svc', svc_clf),
        ('knn', knn_clf)
    ],
    final_estimator=LogisticRegression(),
    cv=5
)

stack_clf.fit(X_train, y_train)

y_pred = stack_clf.predict(X_test)

acc_stack = accuracy_score(y_test, y_pred)

print("Stacking Accuracy:", acc_stack)

Stacking Accuracy: 0.928


Stacking Accuracy: 0.928

### Q7(b) Answer

The stacking classifier achieved an accuracy of **0.928**, which is higher than both hard voting (**0.896**) and soft voting (**0.912**).

This shows that stacking outperforms voting methods in this case.

The main reason is that voting classifiers simply combine predictions by majority vote (hard voting) or by averaging probabilities (soft voting), without learning how to combine them optimally.

In contrast, stacking uses a meta-learner (logistic regression in this case) to learn the best way to combine the predictions of base models. This allows stacking to capture patterns and relationships between the models, leading to better overall performance.


In [13]:
# Q8 - Model Comparison (AUTO)

results = {
    "Decision Tree": acc_dt,
    "Hard Voting": acc_hard,
    "Soft Voting": acc_soft,
    "Bagging": acc_bag,
    "Random Forest": acc_rf,
    "Gradient Boosting (100,0.1)": acc_gb1,
    "Gradient Boosting (300,0.01)": acc_gb2,
    "Stacking": acc_stack
}

# Print all results
print("=== Model Accuracies ===")
for model, acc in results.items():
    print(f"{model}: {acc:.4f}")

# Best model
best_model = max(results, key=results.get)
best_score = results[best_model]

print("\nBest Model:", best_model)
print("Best Accuracy:", round(best_score, 4))

=== Model Accuracies ===
Decision Tree: 0.8560
Hard Voting: 0.8960
Soft Voting: 0.9120
Bagging: 0.9200
Random Forest: 0.9280
Gradient Boosting (100,0.1): 0.8960
Gradient Boosting (300,0.01): 0.9120
Stacking: 0.9280

Best Model: Random Forest
Best Accuracy: 0.928


### Q8 Answer

(a) The best performing models are **Random Forest** and **Stacking**, both achieving an accuracy of **0.928**.  
Random Forest performs very well because it reduces variance by averaging many decision trees while also introducing feature randomness.  
Stacking achieves similar performance by learning how to optimally combine multiple different models.

---

(b) The method that reduces variance the most is **Bagging** (and also Random Forest as an extension of it), since it averages predictions from multiple models trained on different subsets of data.

---

(c) The model that has the highest risk of overfitting is the **Decision Tree**, because it can grow very deep and memorize the training data.  
This is also reflected in its lower test accuracy compared to ensemble methods.

---

(d) Boosting is preferred over bagging when the model suffers from high bias and we want to improve performance by sequentially correcting errors.  
Although in this experiment boosting did not outperform Random Forest, it still provides competitive performance.
